x-pack/filebeat/input/httpjson/request.go

```diff
@@ -39,6 +39,11 @@ type idsStruct struct {
 	err error
 }
 
+type processResponse struct {
+	httpResponse       *http.Response
+	startSignalChannel chan struct{}
+}
```

Comment:
It would be good to have an explanation for why we are doing this. Particularly what the invariants are that `startSignalChannel` is ensuring are - a name change could help this, suggest `idsReady` or similar. I'm wondering also if you can't get away with a lock here (essentially what you are using is a waitgroup, but with only one member).
```diff
diff --git a/x-pack/filebeat/input/httpjson/request.go b/x-pack/filebeat/input/httpjson/request.go
index 49d08ca039..c7aad71c07 100644
--- a/x-pack/filebeat/input/httpjson/request.go
+++ b/x-pack/filebeat/input/httpjson/request.go
@@ -13,6 +13,7 @@ import (
        "net/http"
        "net/url"
        "strings"
+       "sync"
 
        "github.com/PaesslerAG/jsonpath"
 
@@ -41,7 +42,35 @@ type idsStruct struct {
 
 type processResponse struct {
        httpResponse       *http.Response
-       startSignalChannel chan struct{}
+       startSignalChannel *sync.Mutex
+}
+
+func block(resp *http.Response) processResponse {
+       var mu sync.Mutex
+       mu.Lock()
+       return processResponse{resp, &mu}
+}
+
+// On a pointer to get a side-effect. All the methods should agree, but this
+// is just demonstrational.
+func (p *processResponse) block() processResponse {
+       var mu sync.Mutex
+       mu.Lock()
+       p.startSignalChannel = &mu
+       return *p
+}
+
+func (p processResponse) done() {
+       if p.startSignalChannel != nil {
+               p.startSignalChannel.Unlock()
+       }
+}
+
+func (p processResponse) wait() {
+       if p.startSignalChannel != nil {
+               // This abuses sync.Mutex to behave as a waitgroup of one.
+               p.startSignalChannel.Lock()
+       }
 }
 
 func (c *httpClient) do(stdCtx context.Context, req *http.Request) (*http.Response, error) {
@@ -382,10 +411,7 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
                        }
 
                        // passing signal channel here to avoid closing of nil channel in getIdsFromResponses function
-                       intermediateResponseChannel <- processResponse{
-                               httpResponse:       httpResp,
-                               startSignalChannel: make(chan struct{}),
-                       }
+                       intermediateResponseChannel <- block(httpResp)
                        close(intermediateResponseChannel)
                        idsChannel := make(chan idsStruct, 1)
                        go r.getIdsFromResponses(intermediateResponseChannel, idsChannel, r.requestFactories[i+1].replace)
@@ -474,8 +500,7 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
                                resp := processResponse{httpResponse: httpResp}
                                // store data according to response type
                                if !(i == len(r.requestFactories)-1 && len(ids) != 0) {
-                                       resp.startSignalChannel = make(chan struct{})
-                                       intermediateResponseChannel <- resp
+                                       intermediateResponseChannel <- resp.block()
                                }
                                responseChannel <- resp
                        }
@@ -516,7 +541,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                        b, err = io.ReadAll(resp.httpResponse.Body)
                        if err != nil {
                                outputChannel <- idsStruct{err: fmt.Errorf("error while reading response body: %w", err)}
-                               close(resp.startSignalChannel)
+                               resp.done()
                                closeSignalChannels(intermediateResps)
                                return
                        }
@@ -525,7 +550,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                err = resp.httpResponse.Body.Close()
                if err != nil {
                        outputChannel <- idsStruct{err: fmt.Errorf("error closing response body: %w", err)}
-                       close(resp.startSignalChannel)
+                       resp.done()
                        closeSignalChannels(intermediateResps)
                        return
                }
@@ -534,14 +559,14 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                var v interface{}
                if err := json.Unmarshal(b, &v); err != nil {
                        outputChannel <- idsStruct{err: fmt.Errorf("cannot unmarshal data: %w", err)}
-                       close(resp.startSignalChannel)
+                       resp.done()
                        closeSignalChannels(intermediateResps)
                        return
                }
                values, err := jsonpath.Get(replace, v)
                if err != nil {
                        outputChannel <- idsStruct{err: fmt.Errorf("error while getting keys: %w", err)}
-                       close(resp.startSignalChannel)
+                       resp.done()
                        closeSignalChannels(intermediateResps)
                        return
                }
@@ -554,7 +579,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                                        ids = append(ids, fmt.Sprintf("%v", v))
                                default:
                                        r.log.Errorf("events must a number or string, but got %T: skipping", v)
-                                       close(resp.startSignalChannel)
+                                       resp.done()
                                        continue
                                }
                        }
@@ -563,7 +588,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                default:
                        r.log.Errorf("cannot collect IDs from type '%T' : '%v'", values, values)
                }
-               close(resp.startSignalChannel)
+               resp.done()
        }
        outputChannel <- idsStruct{ids: ids}
 }
@@ -661,10 +686,7 @@ func (r *requester) processChainPaginationEvents(stdCtx context.Context, trCtx *
        idsChannel := make(chan idsStruct, 1)
 
        // passing signal channel here to avoid closing of nil channel in getIdsFromResponses function
-       intermediateResponseChannel <- processResponse{
-               httpResponse:       response,
-               startSignalChannel: make(chan struct{}),
-       }
+       intermediateResponseChannel <- block(response)
        close(intermediateResponseChannel)
 
        go r.getIdsFromResponses(intermediateResponseChannel, idsChannel, r.requestFactories[chainIndex].replace)
@@ -742,8 +764,7 @@ func (r *requester) processChainPaginationEvents(stdCtx context.Context, trCtx *
 
                        resp := processResponse{httpResponse: httpResp}
                        if !(i == len(r.requestFactories)-1 && len(ids) != 0) {
-                               resp.startSignalChannel = make(chan struct{})
-                               intermediateResponseChannel <- resp
+                               intermediateResponseChannel <- resp.block()
                        }
                        processingChannel <- resp
                }
@@ -818,6 +839,6 @@ func closeChannels(channels ...chan processResponse) {
 
 func closeSignalChannels(channel <-chan processResponse) {
        for c := range channel {
-               close(c.startSignalChannel)
+               c.done()
        }
 }
diff --git a/x-pack/filebeat/input/httpjson/response.go b/x-pack/filebeat/input/httpjson/response.go
index 7b800a0162..84d8c86813 100644
--- a/x-pack/filebeat/input/httpjson/response.go
+++ b/x-pack/filebeat/input/httpjson/response.go
@@ -136,9 +136,7 @@ func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *tran
 
        for resp := range resps {
                // in case of intermediate responses wait for IDs to be collected
-               if resp.startSignalChannel != nil {
-                       <-resp.startSignalChannel
-               }
+               resp.wait()
 
                iter := rp.pagination.newPageIterator(stdCtx, trCtx, resp.httpResponse)
                for {
```

Reply:
While testing we found that the intermediate response was getting published instead of final response while using double chain calls.

**Cause**: 
HTTP response body should be read first by getIdsFromResponses function and then by startProcessing function. So, in concurrent execution of both of these function the response body may be read by startProcessing function first.

**Solution**:
In startProcessing function wait for getIdsFromResponses function to finish collecting IDs from the response.

I'll update the PR with these changes.

Reply:
The code above is not production-ready, so if you use it, please clean it up.

Reply:
Thanks for your suggestions. I have updated the waiting logic to use mutex instead of channel.